### DISCUSSION TITLE
Cannot use rdev to capture events in the application window.

### DISCUSSION THREAD

**Discussion by NeserCode** [tauri-apps/tauri#7752] [Q&A] [answered]
I am trying to capture keyboard events in the app window using rdev crate, but I am not seeing debugging output for keyboard events inside the app. Debugging output for those keyboard events can be seen elsewhere, on the desktop, in the browser, and even in the app developer's tools. What should I do to get rdev to capture keyboard events in my application, or what data should I provide?

**Comment by FabianLars** [comment, marked as answer]
I assume you're on Windows? And with "events can be seen elsewhere" you're talking about which window has the focus right? In this case i think you need to use this api https://docs.rs/tauri/latest/tauri/struct.Builder.html#method.device_event_filter to change the filter (not sure which filter you need, so you'll have to try them yourself).

**Reply by NeserCode** [reply]
No, they don't work at all.

**Reply by NeserCode** [reply]
This is my main.rs main fn.
```
fn main() {
    let context = tauri::generate_context!();

    tauri::Builder::default()
        .setup(|_app| {
            fn callback(event: Event) {
                match event.event_type {
                    EventType::KeyPress(_) => {
                        println!("KeyPress {:?}", event);
                    }
                    EventType::KeyRelease(_) => {
                        println!("KeyRelease {:?}", event)
                    }
                    _ => (),
                }
            }

            tauri::async_runtime::spawn(async move {
                if let Err(error) = listen(callback) {
                    println!("Error: {:?}", error)
                }
            });

            Ok(())
        })
        .invoke_handler(tauri::generate_handler![use_devtools])
        .device_event_filter(tauri::DeviceEventFilter::Never)
        .run(context)
        .expect("error while running application");
}
```

**Reply by NeserCode** [reply]
OK, it works, maybe i make a mistake in somewhere...

**Comment by HuakunShen** [comment]
@NeserCode which version of rdev are you using?
For me, the mouse moves can be detected, but the entire program terminates when any key is pressed.

Note: if I run the listener without `tauri::async_runtime::spawn`, key listener will work.

**Reply by NeserCode** [reply]
First, it's right that using async thread to contain the listeners. And the way to how to do the same handler in the application is using the `device_events_filter` with "Never" option, which may let the outer api can be detected in your app.
Like this line in your main.rs:
```rust
.device_event_filter(tauri::DeviceEventFilter::Never)
```

**Reply by HuakunShen** [reply]
@NeserCode Thanks, but this didn't work, the app still quits on key press. I noticed that this only happens on MacOS. Windows and Linux work fine.

I made another discussion with more details about this https://github.com/tauri-apps/tauri/discussions/7839.

Have you tried on MacOS?

**Reply by NeserCode** [reply]
Emmm, i think you will find the [Platform-specific](https://docs.rs/tauri/latest/tauri/struct.Builder.html#platform-specific-1) in [mentioned api](https://docs.rs/tauri/latest/tauri/struct.Builder.html#method.device_event_filter).
There shows **Linux / macOS / iOS / Android**: Unsupported.
So, i think this solution can not be worked successfully on these platform. Maybe you should wait changes on your platform or use other solution.
